### PR TITLE
[#5647] improvement(build): update the jar name to add gravitino prefix 

### PR DESCRIPTION
**Pull request by orenccl**
### What changes were proposed in this pull request?

As title

### Why are the changes needed?

Fix: #5647

### Does this PR introduce _any_ user-facing change?

 Makes users easy to know it is from Gravitino

### How was this patch tested?
Execute Gradle command `./gradlew publishToMavenLocal -x test` then check in Maven local path.

### CODE REVIEW

**Comment by orenccl** [comment]
I also encountered this network error on other pull requests. Is there a way to improve this?

```
> Task :clients:client-python:downloadHadoopPack FAILED
* What went wrong:
Execution failed for task ':clients:client-python:downloadHadoopPack'.
> A failure occurred while executing de.undercouch.gradle.tasks.download.internal.DefaultWorkerExecutorHelper$DefaultWorkAction
   > java.net.SocketException: Network is unreachable
```

It seems to be a network issue in the CI environment.

**Comment by jerryshao** [comment]
It is occasional, maybe you can configuration that gradle plugin to increase the timeout or add retry times (but it depends on if that plugin has such configurations).

File: api/build.gradle.kts [hunk 1]
```diff
@@ -35,3 +37,15 @@ dependencies {
...
+tasks.jar {
+  archiveBaseName.set(artifactName)
+}
+
+publishing {
+  publications {
+    withType<MavenPublication>().configureEach {
+      artifactId = artifactName
+    }
+  }
```
Comment (jerryshao) [inline api/build.gradle.kts:50]: Do we need to configure this one by one module, or we can achieve this in the main `build.gradle.kts`?

Comment (jerryshao) [reply api/build.gradle.kts:50]: I think it is achievable, you can implement this in `subprojects`, can you take a try?

Comment (orenccl) [reply api/build.gradle.kts:50]: I'll take a look to see if this can be configured in the main `build.gradle.kts`.

**Comment by orenccl** [comment]
I've discovered that the `artifactId` is the `jarName` without the version.  
(Reference: [Maven Naming Conventions](https://maven.apache.org/guides/mini/guide-naming-conventions.html)).

Additionally, the `artifactId` defaults to `project.name`.  
(Reference: [Gradle Documentation for MavenPublication](https://docs.gradle.org/current/dsl/org.gradle.api.publish.maven.MavenPublication.html)).

Therefore, setting the `artifactId` only once in the root `build.gradle.kts` should resolve the issue.

**Comment by jerryshao** [comment]
For some modules like spark, flink, we already added "gravitino-" as the project name, can you please check if your change will add the prefix again, also please align these modules with other modules.

**Comment by orenccl** [comment]
After looking into it, I found that the `artifactId` set in the root project doesn’t apply if the child `build.gradle.kts` files have their own `artifactId` settings.

Moreover, it also looks like they need to do some custom work to get their naming conventions right.

For example
```kotlin
val scalaVersion: String = project.properties["scalaVersion"] as? String ?: extra["defaultScalaVersion"].toString()
val sparkVersion: String = libs.versions.spark33.get()
val sparkMajorVersion: String = sparkVersion.substringBeforeLast(".")
val baseName = "${rootProject.name}-spark-connector-runtime-${sparkMajorVersion}_$scalaVersion"
```

image

File: build.gradle.kts [hunk 2]
```diff
@@ -413,6 +413,8 @@ subprojects {
           artifact(javadocJar)
         }
 
+        artifactId = "${rootProject.name.lowercase(Locale.getDefault())}-${project.name}"
```
Comment (tengqm) [inline build.gradle.kts:416]: Why the invocation to `Locale.getDefault()`?

Comment (orenccl) [reply build.gradle.kts:416]: I was following the code style in this script, where `toLowerCase(Locale.getDefault())` is actually just `toLowerCase()`.

**Comment by jerryshao** [comment]
I tested locally, LGTM, @orenccl do you have anything else needs to be addressed?

**Comment by orenccl** [comment]
Hi, @jerryshao 
I believe this is good to go, but happy to address any further comments.

File: build.gradle.kts (hunk 2 above)
Comment (jerryshao) [reply build.gradle.kts:416]: Specifying `Local` for `lowercase ()` API is a best practice in JVM language, otherwise we will get an expected result.

**Comment by tengqm** [comment]
> @tengqm do you want to take another look?

Good to learn this convention.
LGTM

**Review by jerryshao — APPROVED** [review]

**Comment by jerryshao** [comment]
> > > @tengqm do you want to take another look?
> > 
> > 
> > Good to learn this convention. LGTM
> 
> p.s. I raised the question because I was checking the use case here. It is not about converting a general text message which are supposed to be localized. My understanding was that JAR names are always ASCII printable characters and we don't need (want to) over-engineer this to consider file names with "イ" or "い".

I see. Yes, it depends. But some static check tools like error-prone, or IDEA itself may complain about this if the `Locale` is not set. So typically, we will add the Locale to the methods like this.

**Comment by jerryshao** [comment]
Hi @tengqm I'm going to merge this if there's no further comment.

**Pull request merged**